[cxx-interop] Reference to type metadata pointer for imported c++ class can be improperly referenced in C++ import header

### Description

Not sure how to reduce this one, my college try of ~1h didn't seem to reproduce the issue.

As noted in https://github.com/swiftlang/swift/issues/80291#issuecomment-2778843342, When the issue with the import header is worked around with `-enable-library-evolution -enable-experimental-feature AssumeResilientCxxTypes`, a new issue showed up

### Reproduction

```
git clone git@github.com:ADKaster/ladybird-browser.git
cd ladybird-browser
git checkout 683f4ed9c63a9c4af9374d1ef52e3469848efa79
swiftly install --use  main-snapshot-2025-04-02
cmake --preset default \
  -DCMAKE_C_COMPILER=$(swiftly use --print-location)/usr/bin/clang \
  -DCMAKE_CXX_COMPILER=$(swiftly use --print-location)/usr/bin/clang++ \
  -DENABLE_SWIFT=ON
ninja -C Build/release LibWeb
```

```
[3273/3377] Building CXX object Lagom/Libraries/LibWeb/CMakeFiles/LibWeb.dir/HTML/Parser/HTMLParser.cpp.o
FAILED: Lagom/Libraries/LibWeb/CMakeFiles/LibWeb.dir/HTML/Parser/HTMLParser.cpp.o
/usr/bin/ccache /home/andrew/.local/share/swiftly/toolchains/main-snapshot-2025-04-02/usr/bin/clang++ -DENABLE_COMPILETIME_FORMAT_CHECK -DLIBWEB_USE_SWIFT -DLibWeb_EXPORTS -DSK_CODEC_DECODES_BMP -DSK_CODEC_DECODES_GIF -DSK_CODEC_DECODES_ICO -DSK_CODEC_DECODES_JPEG -DSK_CODEC_DECODES_PNG -DSK_CODEC_DECODES_RAW -DSK_CODEC_DECODES_WBMP -DSK_CODEC_DECODES_WEBP -DSK_DISABLE_TRACING -DSK_ENABLE_AVX512_OPTS -DSK_ENABLE_PRECOMPILE -DSK_FONTMGR_ANDROID_AVAILABLE -DSK_FONTMGR_FONTCONFIG_AVAILABLE -DSK_FONTMGR_FREETYPE_DIRECTORY_AVAILABLE -DSK_FONTMGR_FREETYPE_EMBEDDED_AVAILABLE -DSK_FONTMGR_FREETYPE_EMPTY_AVAILABLE -DSK_GAMMA_APPLY_TO_A8 -DSK_GANESH -DSK_GL -DSK_HAS_WUFFS_LIBRARY -DSK_R32_SHIFT=16 -DSK_SUPPORT_PDF -DSK_TYPEFACE_FACTORY_FREETYPE -DSK_USE_PERFETTO -DSK_USE_VMA -DSK_VULKAN -DSK_XML -DUSE_FONTCONFIG=1 -DUSE_VULKAN=1 -D_FILE_OFFSET_BITS=64 -I/home/andrew/ladybird-org/ladybird-browser -I/home/andrew/ladybird-org/ladybird-browser/Services -I/home/andrew/ladybird-org/ladybird-browser/Libraries -I/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom -I/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom/Services -I/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom/Libraries -I/home/andrew/.local/share/swiftly/toolchains/main-snapshot-2025-04-02/usr/include -I/home/andrew/ladybird-org/ladybird-browser/Meta/Lagom/../.. -I/home/andrew/ladybird-org/ladybird-browser/Meta/Lagom/../../Libraries -I/home/andrew/ladybird-org/ladybird-browser/Meta/Lagom/../../Services -I/home/andrew/ladybird-org/ladybird-browser/Build/release -I/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom/Libraries/LibWeb/include -I/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom/AK/include -I/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom/Libraries/LibCore/include -I/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom/Libraries/LibGfx/include -I/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom/Libraries/LibGC/include -isystem /home/andrew/ladybird-org/ladybird-browser/Build/release/vcpkg_installed/x64-linux-dynamic/include -isystem /home/andrew/ladybird-org/ladybird-browser/Build/release/vcpkg_installed/x64-linux-dynamic/include/skia -O2 -g -DNDEBUG -std=c++23 -fPIC -fcolor-diagnostics -Wall -Wextra -fno-exceptions -ffp-contract=off -Wcast-qual -Wformat=2 -Wimplicit-fallthrough -Wlogical-op -Wmissing-declarations -Wmissing-field-initializers -Wsuggest-override -Wno-invalid-offsetof -Wno-unknown-warning-option -Wno-unused-command-line-argument -Werror -fconstexpr-steps=16777216 -Wmissing-prototypes -Wno-implicit-const-int-float-conversion -Wno-user-defined-literals -Wno-unqualified-std-cast-call -fno-semantic-interposition -fvisibility-inlines-hidden -fstack-protector-strong -fstrict-flex-arrays=2 -Wno-maybe-uninitialized -Wno-shorten-64-to-32 -fsigned-char -ggnu-pubnames -fPIC -O2 -g1 -Wno-overloaded-virtual -MD -MT Lagom/Libraries/LibWeb/CMakeFiles/LibWeb.dir/HTML/Parser/HTMLParser.cpp.o -MF Lagom/Libraries/LibWeb/CMakeFiles/LibWeb.dir/HTML/Parser/HTMLParser.cpp.o.d -o Lagom/Libraries/LibWeb/CMakeFiles/LibWeb.dir/HTML/Parser/HTMLParser.cpp.o -c /home/andrew/ladybird-org/ladybird-browser/Libraries/LibWeb/HTML/Parser/HTMLParser.cpp
In file included from /home/andrew/ladybird-org/ladybird-browser/Libraries/LibWeb/HTML/Parser/HTMLParser.cpp:53:
/home/andrew/ladybird-org/ladybird-browser/Build/release/Lagom/Libraries/LibWeb/include/LibWeb-Swift.h:5857:12: error: use of undeclared identifier '__CellNested'
 5857 |     return __CellNested::_impl::$sSo2GCO4CellV7VisitorVMa(0)._0;
      |            ^
1 error generated.
```

`$sSo2GCO4CellV7VisitorVMa` corresponds to 
```
$ ~/demumble/demumble \$sSo2GCO4CellV7VisitorVMa
type metadata accessor for __C.GC.Cell.Visitor
```

### Expected behavior

Build with no warnings or errors

### Environment

Swift version 6.2-dev (LLVM 86709865028dc89, Swift 2e6d7f71608bbad)
Target: x86_64-unknown-linux-gnu
Build config: +assertions

swiftly main-snapshot-2025-04-02 on Ubuntu 24.04

### Additional information

_No response_

--- a/include/swift/ClangImporter/ClangImporter.h
+++ b/include/swift/ClangImporter/ClangImporter.h
@@ -773,6 +773,9 @@ getPrivateFileIDAttrs(const clang::CXXRecordDecl *decl);
 ///
 /// Returns false if \a decl was not imported by ClangImporter.
 bool declIsCxxOnly(const Decl *decl);
+
+/// Is this DeclContext an `enum` that represents a C++ namespace?
+bool isClangNamespace(const DeclContext *dc);
 } // namespace importer
 
 struct ClangInvocationFileMapping {

--- a/lib/ClangImporter/ClangImporter.cpp
+++ b/lib/ClangImporter/ClangImporter.cpp
@@ -8793,3 +8793,10 @@ bool importer::declIsCxxOnly(const Decl *decl) {
   }
   return false;
 }
+
+bool importer::isClangNamespace(const DeclContext *dc) {
+  if (const auto *ed = dc->getSelfEnumDecl())
+    return isa_and_nonnull<clang::NamespaceDecl>(ed->getClangDecl());
+
+  return false;
+}

--- a/lib/ClangImporter/ImportDecl.cpp
+++ b/lib/ClangImporter/ImportDecl.cpp
@@ -75,6 +75,7 @@
 #include "llvm/ADT/StringMap.h"
 #include "llvm/ADT/StringSwitch.h"
 #include "llvm/ADT/TinyPtrVector.h"
+#include "llvm/Support/Casting.h"
 #include "llvm/Support/Path.h"
 
 #include <algorithm>
@@ -3728,13 +3729,6 @@ namespace {
           return attr;
       }
       return nullptr;
-    }
-
-    static bool isClangNamespace(const DeclContext *dc) {
-      if (const auto *ed = dc->getSelfEnumDecl())
-        return isa<clang::NamespaceDecl>(ed->getClangDecl());
-
-      return false;
     }
 
     Decl *importFunctionDecl(

--- a/lib/PrintAsClang/ClangSyntaxPrinter.cpp
+++ b/lib/PrintAsClang/ClangSyntaxPrinter.cpp
@@ -207,7 +207,8 @@ void ClangSyntaxPrinter::printNamespace(
 void ClangSyntaxPrinter::printParentNamespaceForNestedTypes(
     const ValueDecl *D, llvm::function_ref<void(raw_ostream &OS)> bodyPrinter,
     NamespaceTrivia trivia) const {
-  if (!isa_and_nonnull<NominalTypeDecl>(D->getDeclContext()->getAsDecl())) {
+  if (!isa_and_nonnull<NominalTypeDecl>(D->getDeclContext()->getAsDecl()) ||
+      importer::isClangNamespace(D->getDeclContext())) {
     bodyPrinter(os);
     return;
   }

--- a/lib/PrintAsClang/PrintClangValueType.cpp
+++ b/lib/PrintAsClang/PrintClangValueType.cpp
@@ -630,18 +630,21 @@ void ClangValueTypePrinter::printTypeGenericTraits(
     bool isOpaqueLayout) {
   auto *NTD = dyn_cast<NominalTypeDecl>(typeDecl);
   ClangSyntaxPrinter printer(typeDecl->getASTContext(), os);
+  if (typeDecl->hasClangNode()) {
+    /// Print a reference to the type metadata function for a C++ type.
+    printer.printParentNamespaceForNestedTypes(typeDecl, [&](raw_ostream &os) {
+      printer.printNamespace(
+          cxx_synthesis::getCxxImplNamespaceName(), [&](raw_ostream &os) {
+            ClangSyntaxPrinter(typeDecl->getASTContext(), os)
+                .printCTypeMetadataTypeFunction(typeDecl, typeMetadataFuncName,
+                                                typeMetadataFuncRequirements);
+          });
+    });
+  }
+
   // FIXME: avoid popping out of the module's namespace here.
   os << "} // end namespace \n\n";
   os << "namespace swift SWIFT_PRIVATE_ATTR {\n";
-
-  if (typeDecl->hasClangNode()) {
-    /// Print a reference to the type metadata function for a C++ type.
-    ClangSyntaxPrinter(typeDecl->getASTContext(), os).printNamespace(
-        cxx_synthesis::getCxxImplNamespaceName(), [&](raw_ostream &os) {
-          ClangSyntaxPrinter(typeDecl->getASTContext(), os).printCTypeMetadataTypeFunction(
-              typeDecl, typeMetadataFuncName, typeMetadataFuncRequirements);
-        });
-  }
   auto classDecl = dyn_cast<ClassDecl>(typeDecl);
   bool addPointer =
       typeDecl->isObjC() || (classDecl && classDecl->isForeignReferenceType());
@@ -676,10 +679,11 @@ void ClangValueTypePrinter::printTypeGenericTraits(
   ClangSyntaxPrinter(typeDecl->getASTContext(), os).printInlineForHelperFunction();
   os << "void * _Nonnull getTypeMetadata() {\n";
   os << "    return ";
-  if (!typeDecl->hasClangNode()) {
+  if (typeDecl->hasClangNode())
+    printer.printBaseName(moduleContext);
+  else
     printer.printBaseName(typeDecl->getModuleContext());
-    os << "::";
-  }
+  os << "::";
   if (!printer.printNestedTypeNamespaceQualifiers(typeDecl))
     os << "::";
   os << cxx_synthesis::getCxxImplNamespaceName() << "::";

--- a/test/Interop/CxxToSwiftToCxx/bridge-cxx-struct-back-to-cxx.swift
+++ b/test/Interop/CxxToSwiftToCxx/bridge-cxx-struct-back-to-cxx.swift
@@ -200,10 +200,7 @@ public struct Strct {
 // CHECK:  ns::ImmortalTemplate<int> *_Nonnull retImmortalTemplate() noexcept SWIFT_SYMBOL({{.*}}) SWIFT_WARN_UNUSED_RESULT {
 // CHECK-NEXT: return UseCxxTy::_impl::$s8UseCxxTy19retImmortalTemplateSo2nsO0028ImmortalTemplateCInt_jBAGgnbVyF();
 // CHECK-NEXT: }
-
-// CHECK: } // end namespace
 // CHECK-EMPTY:
-// CHECK-NEXT: namespace swift SWIFT_PRIVATE_ATTR {
 // CHECK-NEXT: namespace _impl {
 // CHECK-EMPTY:
 // CHECK-NEXT: // Type metadata accessor for NonTrivialTemplateInt
@@ -212,6 +209,9 @@ public struct Strct {
 // CHECK-EMPTY:
 // CHECK-NEXT: } // namespace _impl
 // CHECK-EMPTY:
+// CHECK-NEXT: } // end namespace
+// CHECK-EMPTY:
+// CHECK-NEXT: namespace swift SWIFT_PRIVATE_ATTR {
 // CHECK-NEXT: #pragma clang diagnostic push
 // CHECK-NEXT: #pragma clang diagnostic ignored "-Wc++17-extensions"
 // CHECK-NEXT: template<>
@@ -219,7 +219,7 @@ public struct Strct {
 // CHECK-NEXT: template<>
 // CHECK-NEXT: struct TypeMetadataTrait<ns::NonTrivialTemplateInt> {
 // CHECK-NEXT:   static SWIFT_INLINE_PRIVATE_HELPER void * _Nonnull getTypeMetadata() {
-// CHECK-NEXT:     return _impl::$sSo2nsO0030NonTrivialTemplateCInt_hHAFhrbVMa(0)._0;
+// CHECK-NEXT:     return UseCxxTy::_impl::$sSo2nsO0030NonTrivialTemplateCInt_hHAFhrbVMa(0)._0;
 // CHECK-NEXT:   }
 // CHECK-NEXT: };
 // CHECK-NEXT: namespace _impl{
@@ -240,9 +240,6 @@ public struct Strct {
 // CHECK-NEXT: return result;
 // CHECK-NEXT: }
 // CHECK-EMPTY:
-// CHECK-NEXT: } // end namespace
-// CHECK-EMPTY:
-// CHECK-NEXT: namespace swift SWIFT_PRIVATE_ATTR {
 // CHECK-NEXT: namespace _impl {
 // CHECK-EMPTY:
 // CHECK-NEXT: // Type metadata accessor for NonTrivialTemplateTrivial
@@ -251,6 +248,9 @@ public struct Strct {
 // CHECK-EMPTY:
 // CHECK-NEXT: } // namespace _impl
 // CHECK-EMPTY:
+// CHECK-NEXT: } // end namespace
+// CHECK-EMPTY:
+// CHECK-NEXT: namespace swift SWIFT_PRIVATE_ATTR {
 // CHECK-NEXT: #pragma clang diagnostic push
 // CHECK-NEXT: #pragma clang diagnostic ignored "-Wc++17-extensions"
 // CHECK-NEXT: template<>
@@ -258,7 +258,7 @@ public struct Strct {
 // CHECK-NEXT: template<>
 // CHECK-NEXT: struct TypeMetadataTrait<ns::NonTrivialTemplateTrivial> {
 // CHECK-NEXT:   static SWIFT_INLINE_PRIVATE_HELPER void * _Nonnull getTypeMetadata() {
-// CHECK-NEXT:     return _impl::$sSo2nsO0042NonTrivialTemplatensTrivialinNS_HlGFlenawcVMa(0)._0;
+// CHECK-NEXT:     return UseCxxTy::_impl::$sSo2nsO0042NonTrivialTemplatensTrivialinNS_HlGFlenawcVMa(0)._0;
 // CHECK-NEXT:   }
 // CHECK-NEXT: };
 // CHECK-NEXT: namespace _impl{

--- a/test/Interop/CxxToSwiftToCxx/nested-cxx-class-back-to-swift.swift
+++ b/test/Interop/CxxToSwiftToCxx/nested-cxx-class-back-to-swift.swift
@@ -1,0 +1,27 @@
+// RUN: %empty-directory(%t)
+// RUN: split-file %s %t
+
+// RUN: %target-swift-frontend %t/use-cxx-types.swift -module-name UseCxxTy -typecheck -verify -emit-clang-header-path %t/UseCxxTy.h -I %t -enable-experimental-cxx-interop -clang-header-expose-decls=all-public -disable-availability-checking
+// RUN: cat %t/header.h >> %t/full-header.h
+// RUN: cat %t/UseCxxTy.h >> %t/full-header.h
+// RUN: %target-interop-build-clangxx -std=c++20 -c -xc++-header %t/full-header.h -o %t/o.o
+
+//--- header.h
+
+struct Cell { class Visitor {}; };
+
+//--- module.modulemap
+module CxxTest {
+    header "header.h"
+    requires cplusplus
+}
+
+//--- use-cxx-types.swift
+import CxxTest
+
+public extension Cell.Visitor {
+    func visit() {}
+}
+
+public func f() -> [Cell.Visitor] {
+}

--- a/test/Interop/ObjCToSwiftToObjCxx/bridge-objc-types-back-to-objcxx.swift
+++ b/test/Interop/ObjCToSwiftToObjCxx/bridge-objc-types-back-to-objcxx.swift
@@ -94,7 +94,7 @@ public func retObjCClassArray() -> [ObjCKlass] {
 // CHECK-NEXT: template<>
 // CHECK-NEXT: struct TypeMetadataTrait<ObjCKlass*> {
 // CHECK-NEXT:   static SWIFT_INLINE_PRIVATE_HELPER void * _Nonnull getTypeMetadata() {
-// CHECK-NEXT:     return _impl::$sSo9ObjCKlassCMa(0)._0;
+// CHECK-NEXT:     return UseObjCTy::_impl::$sSo9ObjCKlassCMa(0)._0;
 // CHECK-NEXT:   }
 // CHECK-NEXT: };
 

--- a/test/Interop/SwiftToCxx/structs/nested-structs-in-cxx.swift
+++ b/test/Interop/SwiftToCxx/structs/nested-structs-in-cxx.swift
@@ -68,3 +68,11 @@ public class TestObject {
         case invalid
     }
 }
+
+extension RecordConfig.File {
+    public func getFileExtension() -> String { ".wav" }
+}
+
+public func getFiles() -> [RecordConfig.File] {
+    []
+}

